client/components/tinymce/plugins/embed/dialog.jsx

```diff
@@ -0,0 +1,111 @@
+/**
+ * External dependencies
+ */
+import PropTypes from 'prop-types';
+import React from 'react';
+
+/**
+ * Internal dependencies
+ */
+import Button from 'components/button';
+import Dialog from 'components/dialog';
+import FormTextInput from 'components/forms/form-text-input';
+import { localize } from 'i18n-calypso';
+
+/*
+ * Shows the URL of am embed and allows it to be edited.
+ */
+export class EmbedDialog extends React.Component {
+	static propTypes = {
+		embedUrl: PropTypes.string,
+		isVisible: PropTypes.bool,
+
+		// Event handlers
+		onCancel: PropTypes.func.isRequired,
+		onUpdate: PropTypes.func.isRequired,
+
+		// Inherited
+		translate: PropTypes.func.isRequired,
+	};
+
+	static defaultProps = {
+		embedUrl: '',
+		isVisible: false,
+	};
+
+	state = {
+		embedUrl: this.props.embedUrl,
+	};
+
+	/**
+	 * Reset `state.embedUrl` whenever the component's dialog is opened or closed.
+	 *
+	 * If this were not done, then switching back and forth between multiple embeds would result in
+	 * `state.embedUrl` being incorrect. For example, when the second embed was opened,
+	 * `state.embedUrl` would equal the value of the first embed, since it initially set the
+	 * state.
+	 *
+	 * @param {object} nextProps The properties that will be received.
+	 */
+	componentWillReceiveProps = ( nextProps ) => {
+		this.setState( {
+			embedUrl: nextProps.embedUrl,
+		} );
+	};
+
+	onChangeEmbedUrl = ( event ) => {
+		this.setState( {
+			embedUrl: event.target.value,
+		} );
+	};
+
+	onUpdate = () => {
```

Comment:
It may be worth debouncing both of the input events for perf reasons.

Reply:
Did you mean that `onChangeEmbedUrl` (the `<input onChange...` handler) should be debounced, or `onUpdate` and `onCancel`? The latter two only fire once when the `Update` and `Cancel` buttons are clicked, unless I'm missing something.

I'm planning on debouncing part of `onChangeEmbedUrl` in #18171 , because that PR will add a preview embed of the new URL:

https://github.com/Automattic/wp-calypso/pull/18171/files#diff-e88450610ed12ecb3fc80bffa4181e72R65

It doesn't debounce the `setState()` call, though. I would think that would cause problems, since it's a controlled component. Wouldn't it interfere with the user typing a new URL in quickly, or am I misunderstanding how it works? It seems to work ok in my tests, and I don't see any warnings about it when searching, but I just wanna double check.

Reply:
Oops, yes sorry, `onChangeEmbedUrl` :P I'm glad you know which I meant!
I feel like debouncing all of `onChangeEmbedUrl` might be sensible to avoid re-renders on every keypress (since the state is changing).

It would also avoid setting focus on every key up in #18171 - not sure if that helps or hinders that situation though.

Reply:
👍 thanks, I'll go with that approach :)

Reply:
Hmmm, I added the debounce in 2134a11, but that broke the unit tests and introduced a bug where clicking `Update` before the debounce callback has ran will cause the old URL to be returned to TinyMCE instead of the new one.

Those are fixable problems, but I'm just wondering if it's worth making the code more complex to avoid calling `render()` here, especially since I don't think there'll be any diffs between the virtual DOM and the real one, since `FormTextInput` is an uncontrolled component. Also because I think the most common use case will be copy/pasting a URL, rather than manually typing it.

I think _previewing_ the new URL in #18171 should definitely be debounced, but I'm just not sure that `setState` should be.

https://github.com/facebook/react/issues/1360#issuecomment-39680422

Is there a good way to tell what kind of performance impact `setState` will have if left undebounced? Maybe profiling w/ Chrome dev tools or something? I can look into that tomorrow if you think it's worth it.